Separate CLI migration procedure per source provider (#463)
kubev2v/forklift-documentation: documentation/modules/new-migrating-virtual-machines-cli.adoc

+
[source,yaml,subs="attributes+"]
----
$  cat << EOF | {oc} apply -f -
apiVersion: forklift.konveyor.io/v1beta1
kind: NetworkMap
metadata:
  name: <network_map>
  namespace: <namespace>
spec:
  map:
    - destination:
        name: <network_name>
        type: pod <1>
      source:<2>
        id: <source_network_id>
        name: <source_network_name>
    - destination:
        name: <network_attachment_definition> <3>
        namespace: <network_attachment_definition_namespace> <4>
        type: multus
      source:<5>
        id: <source_network_id>
        name: <source_network_name>
  provider:
    source:
      name: <source_provider>
      namespace: <namespace>
    destination:
      name: <destination_provider>
      namespace: <namespace>
EOF
----
<1> Allowed values are `pod` and `multus`.
<2> You can use either the `id` _or_ the `name` parameter to specify the source network. For `id`, specify the {osp} network UUID.
<3> Specify a network attachment definition for each additional {virt} network.
<4> Required only when `type` is `multus`. Specify the namespace of the {virt} network attachment definition.

> <5> Same as `source` above.

Reviewer: missed this one?

Author: Fixed.

endif::[]

ifdef::cnv[]
[start=3]
. Create a `NetworkMap` manifest to map the source and destination networks:
+
[source,yaml,subs="attributes+"]
----
$  cat << EOF | {oc} apply -f -
apiVersion: forklift.konveyor.io/v1beta1
kind: NetworkMap
metadata:
  name: <network_map>
  namespace: <namespace>
spec:
  map:
    - destination:
        name: <network_name>
        type: pod <1>
      source:
        name: <network_name>
        type: pod
    - destination:
        name: <network_attachment_definition> <2>
        namespace: <network_attachment_definition_namespace> <3>
        type: multus
      source:
        name: <network_attachment_definition>
        namespace: <network_attachment_definition_namespace> <4>
        type: multus
  provider:
    source:
      name: <source_provider>
      namespace: <namespace>
    destination:
      name: <destination_provider>
      namespace: <namespace>
EOF
----
<1> Allowed values are `pod` and `multus`.
<2> Specify a network attachment definition for each additional {virt} network.

> <3> Required only when `type` is `multus`. Specify the namespace of the {virt} network attachment definition.

Reviewer: @RichardHoch I meant that the explanation about the 'namespace' can stay but it's still duplicated below, how about changing this to
```suggestion
<3> Required only when `type` is `multus`. The `namespace` can either be specified using the namespace property or with a name built as follows: `<network_namespace>/<network_name>`.
```
and drop `<4>`? (I think there's a confusing about the " or with a name built as follows: `<network_namespace>/<network_name>`" - it also applies here)

Author: Fixed.

<6>  Specify the API endpoint URL of the vCenter or the ESX/ESX, for example, `https://<vCenter_host>/sdk`.
endif::[]
ifdef::rhv[]
[source,yaml,subs="attributes+"]
----
$ cat << EOF | {oc} apply -f -
apiVersion: v1
kind: Secret
metadata:
  name: <secret>
  namespace: <namespace>
  ownerReferences: <1>
    - apiVersion: forklift.konveyor.io/v1beta1
      kind: Provider
      name: <provider_name>
      uid: <provider_uid>
  labels:
    createdForProviderType: ovirt
    createdForResourceType: providers
type: Opaque
stringData:
  user: <user> <2>
  password: <password> <3>
  insecureSkipVerify: <true/false> <4>
  cacert: | <5>
    <ca_certificate>
  url: <api_end_point> <6>
EOF
----
<1> The `ownerReferences` section is optional.
<2> Specify the {rhv-short} {manager} user.
<3> Specify the user password.
<4> Specify `<true>` to skip certificate verification, specify `false` to verify the certificate. Defaults to `false` if not specified. Skipping certificate verification proceeds with an insecure migration and then the certificate is not required. Insecure migration means that the transferred data is sent over an insecure connection and potentially sensitive data could be exposed.

> <5> When this field is not set and 'skip certificate verification' is disabled, {project-short} attempts to use the system CA.

Author: I think i got it...

<10> Optional: You can specify up to two hooks for a VM. Each hook must run during a separate migration step.
<11> Specify the name of the `Hook` CR.
<12> Allowed values are `PreHook`, before the migration plan starts, or `PostHook`, after the migration is complete.
endif::[]

ifdef::rhv[]
[start=6]
. Create a `Plan` manifest for the migration:
+
[source,yaml,subs="attributes+"]
----
$ cat << EOF | {oc} apply -f -
apiVersion: forklift.konveyor.io/v1beta1
kind: Plan
metadata:
  name: <plan> <1>
  namespace: <namespace>
  preserveClusterCpuModel: true <2>
spec:
  warm: true <3>
  provider:
    source:
      name: <source_provider>
      namespace: <namespace>
    destination:
      name: <destination_provider>
      namespace: <namespace>
  map: <4>
    network: <5>
      name: <network_map> <6>
      namespace: <namespace>
    storage: <7>
      name: <storage_map> <8>
      namespace: <namespace>
  targetNamespace: <target_namespace>
  vms: <9>
    - id: <source_vm> <10>
    - name: <source_vm>
      hooks: <11>
        - hook:
            namespace: <namespace>
            name: <hook> <12>
          step: <step> <13>
EOF
----
<1> Specify the name of the `Plan` CR.

> <2> See note below.

Author: @ahadas I didn't upload the note (include it in the file) because I am having a formatting problem with it -- but if you're good with the text, I'll go back to working on its formatting.

> <2> See note below.

Author: Comment added.

<3> Specify whether the migration is warm or cold. If you specify a warm migration without specifying a value for the `cutover` parameter in the `Migration` manifest, only the precopy stage will run.
<4> Specify only one network map and one storage map per plan.
<5> Specify a network mapping even if the VMs to be migrated are not assigned to a network. The mapping can be empty in this case.
<6> Specify the name of the `NetworkMap` CR.
<7> Specify a storage mapping even if the VMs to be migrated are not assigned with disk images. The mapping can be empty in this case.
<8> Specify the name of the `StorageMap` CR.
<9> You can use either the `id` _or_ the `name` parameter to specify the source VMs.
<10> Specify the {rhv-short} VM UUID.
<11> Optional: You can specify up to two hooks for a VM. Each hook must run during a separate migration step.
<12> Specify the name of the `Hook` CR.
<13> Allowed values are `PreHook`, before the migration plan starts, or `PostHook`, after the migration is complete.
endif::[]

> 

Author: @ahadas  I had a foirmatting issue with the note -- Here is my proposed text:

[NOTE]
 ====
* If the migrated machines are set with a custom CPU model, they will be set with that CPU model in the destination cluster, not matter what the setting of `preserveClusterCpuModel` is.

* If the migrated machine are _not_ set with a custom CPU model:

** If `preserveClusterCpuModel` is set to 'true`, {project-short} checks the CPU model of the VM when it runs in {rhv-short}, based on the cluster's configuration, and then sets the migrated VM with that CPU model.
** If `preserveClusterCpuModel` is set to 'false`, {project-short} does not set a CPU type and the VMs are set with the default CPU model of the destination cluster.
====

Reviewer: looks good

+
[source,yaml,subs="attributes+"]
----
$  cat << EOF | {oc} apply -f -
apiVersion: forklift.konveyor.io/v1beta1
kind: NetworkMap
metadata:
  name: <network_map>
  namespace: <namespace>
spec:
  map:
    - destination:
        name: <network_name>
        type: pod <1>
      source: <2>
        id: <source_network_id>
        name: <source_network_name>
    - destination:
        name: <network_attachment_definition> <3>
        namespace: <network_attachment_definition_namespace> <4>
        type: multus
      source: <5>
        id: <source_network_id>
        name: <source_network_name>
  provider:
    source:
      name: <source_provider>
      namespace: <namespace>
    destination:
      name: <destination_provider>
      namespace: <namespace>
EOF
----
<1> Allowed values are `pod` and `multus`.
<2> You can use either the `id` _or_ the `name` parameter to specify the source network. For `id`, specify the VMware vSphere network MOR.
<3> Specify a network attachment definition for each additional {virt} network.
<4> Required only when `type` is `multus`. Specify the namespace of the {virt} network attachment definition.

> <5> Same as `source` above.

Reviewer: this sounds like the 'id' / 'name' property should have the same value as in the mapping above - how about dropping `<5>`? (same for the other provider below)

<3> Specify a network attachment definition for each additional {virt} network.
<4> Required only when `type` is `multus`. Specify the namespace of the {virt} network attachment definition.
<5> Same as `source` above.
endif::[]

ifdef::cnv[]
[start=3]
. Create a `NetworkMap` manifest to map the source and destination networks:
+
[source,yaml,subs="attributes+"]
----
$  cat << EOF | {oc} apply -f -
apiVersion: forklift.konveyor.io/v1beta1
kind: NetworkMap
metadata:
  name: <network_map>
  namespace: <namespace>
spec:
  map:
    - destination:
        name: <network_name>
        type: pod <1>
      source: <2>
        name: <network_name>
        type: pod
    - destination:
        name: <network_attachment_definition> <3>
        namespace: <network_attachment_definition_namespace> <4>
        type: multus
      source:
        name: <network_attachment_definition> <5>
        namespace: <network_attachment_definition_namespace> <6>
        type: multus
  provider:
    source:
      name: <source_provider>
      namespace: <namespace>
    destination:
      name: <destination_provider>
      namespace: <namespace>
EOF
----
<1> Allowed values are `pod` and `multus`.

> <2> You can use either the `id` _or_ the `name` parameter to specify the source network. For `id`, specify a pod network that is similar to the destination.

Reviewer: `<2>` is incorrect for OpenShift provider (there's no `id`), I think we can drop `<2>` from here

<3> Specify a network attachment definition for each additional {virt} network.
<4> Required only when `type` is `multus`. Specify the namespace of the {virt} network attachment definition.
<5> Specify a network attachment definition for each additional {virt} network.

> <6> Required only when `type` is `multus`. Here, `namespace` can either be specified using the namespace property or with a name built as follows: `<network_namespace>/<network_name>`.

Reviewer: no need for `<5>` and `<6>` since they are the same as `<3>` and `<4>`

Author: @ahadas <4> and <6> are not identical. Should I drop <6> , rewrite <4>, or leave as is -- will drop callout <5> in any case.

Reviewer: yeah, the part about the 'namespace' should stay

// Module included in the following assemblies:
//
// * documentation/doc-Migration_Toolkit_for_Virtualization/master.adoc

:_content-type: PROCEDURE
[id="new-migrating-virtual-machines-cli_{context}"]

ifdef::vmware[]
= Migrating a VMware vSphere source provider

> Before you can migrate a VMware vSphere source provider using the CLI, you must obtain the SHA-1 fingerprint of a vCenter host, as described in the procedure that follows.

Author: @ahadas Good to know. In this PR, I will focus only on the CLI side of things and will remove all references to the fingerprint in one or more other PRs.

====
endif::[]
ifdef::cnv[]
= Migrating a Red Hat {virt} source provider

You can use a Red Hat {virt} provider as both a source provider and destination provider.

Specifically, the host cluster that is automatically added as a {virt} provider can be used as both a source provider and a destination provider.

You can use the CLI to migrate VMs from the cluster that {project-short} is deployed on to another cluster, or from a remote cluster to the cluster that {project-short} is deployed on.
endif::[]

.Procedure
. Create a `Secret` manifest for the source provider credentials:
+
ifdef::vmware[]
[source,yaml,subs="attributes+"]
----
$ cat << EOF | {oc} apply -f -
apiVersion: v1
kind: Secret
metadata:
  name: <secret>
  namespace: <namespace>
  ownerReferences: <1>
    - apiVersion: forklift.konveyor.io/v1beta1
      kind: Provider
      name: <provider_name>
      uid: <provider_uid>
  labels:
    createdForProviderType: vsphere
    createdForResourceType: providers
type: Opaque
stringData:
  user: <user> <2>
  password: <password> <3>
  insecureSkipVerify: <true/false> <4>
  url: <api_end_point> <5>

>   thumbprint: <vcenter_fingerprint> <6>

Author: @RichardHoch You stopped here.

> EOF

Author: @ahadas I have copy-pasted the definition for RHV -- Enter the {manager} CA certificate unless it was replaced by a third-party certificate, in which case enter the {manager} Apache CA certificate. You can retrieve the {manager} CA certificate at https://<engine_host>/ovirt-engine/services/pki-resource?resource=ca-certificate&format=X509-PEM-CA.  
Please correct any details (I would guess the certificate address is not correct).

> EOF

Reviewer: ack, I'll comment on that line below

> EOF

Author: @ahadas This is vSphere... would it be:

"Enter the CA certificate for the API server of the vSphere server unless it was replaced by a third-party certificate, in which case enter the {manager} Apache CA certificate."  [Do we know the address of the vSphere server?]

> EOF

Reviewer: I'll comment below

----
<1> The `ownerReferences` section is optional.
<2> Specify the vCenter user.
<3> Specify the user password.

> <4> Specify `<true>` to skip certificate verification, which proceeds with an insecure migration and then the certificate is not required. Insecure migration means that the transferred data is sent over an insecure connection and potentially sensitive data could be exposed. Specifying `<false>` verifies the certificate.

Author: @ahadas This property is not part of the OVA migration YAML. Is that OK? Fixed for the rest.

> <4> Specify `<true>` to skip certificate verification, which proceeds with an insecure migration and then the certificate is not required. Insecure migration means that the transferred data is sent over an insecure connection and potentially sensitive data could be exposed. Specifying `<false>` verifies the certificate.

Reviewer: yes, it's OK

<4> Specify `<true>` to skip certificate verification, which proceeds with an insecure migration and then the certificate is not required. Insecure migration means that the transferred data is sent over an insecure connection and potentially sensitive data could be exposed. Specifying `<false>` verifies the certificate.
<5> Specify the name of the {osp} region.
<6> Enter the CA certificate for connecting to the source environment. The certificate is not used when `insecureSkipVerify` is set to `<true>`. Note that if you set `insecureSkipVerify` to `false` and you do not set  `cacert`, {project-short} attempts to use the system CA.
<7> Specify the API end point URL, for example, `https://<identity_service>/v3`.
endif::[]

ifdef::cnv[]
[source,yaml,subs="attributes+"]
----
$ cat << EOF | {oc} apply -f -
apiVersion: v1
kind: Secret
metadata:
  name: <secret>
  namespace: <namespace>
  ownerReferences: <1>
    - apiVersion: forklift.konveyor.io/v1beta1
      kind: Provider
      name: <provider_name>
      uid: <provider_uid>
  labels:
    createdForProviderType: openshift
    createdForResourceType: providers
type: Opaque
stringData:
  user: <user> <2>
  password: <password> <3>

>   insecureSkipVerify: <true/false> <4>

Reviewer: here we'll also have `cacert` which is similar to that of RHV and OpenStack

Author: Same comment as above -- Here CNV is the source.
Text for callout now reads
 Enter the {manager} CA certificate unless it was replaced by a third-party certificate, in which case enter the {manager} Apache CA certificate. You can retrieve the {manager} CA certificate at https://<engine_host>/ovirt-engine/services/pki-resource?resource=ca-certificate&format=X509-PEM-CA.

Reviewer: here it's a CA certificate for the API server of the OpenShift cluster

Author: @ahadas  Thursday evening: 

1. would that be "Enter the CA certificate for the API server of the {virt} server unless it was replaced by a third-party certificate, in which case enter the {manager} Apache CA certificate."?

2.  Do we need to say anything about the URL for the API server of the OpenShift cluster or is that address known to the user already?

  url: <api_end_point> <5>
EOF
----
<1> The `ownerReferences` section is optional.
<2> Specify the {virt} user.
<3> Specify the user password.
<4> Specify `<true>` to skip certificate verification, which proceeds with an insecure migration and then the certificate is not required. Insecure migration means that the transferred data is sent over an insecure connection and potentially sensitive data could be exposed. Specifying `<false>` verifies the certificate.

> <5> Specify the API end point URL.

Reviewer: we have a chance for being more explicit here and say "API server"

endif::[]
+

. Create a `Provider` manifest for the source provider:
+
ifdef::vmware[]
[source,yaml,subs="attributes+"]
----
$ cat << EOF | {oc} apply -f -
apiVersion: forklift.konveyor.io/v1beta1
kind: Provider
metadata:
  name: <source_provider>
  namespace: <namespace>
spec:

>   type: vmware

Reviewer: ```suggestion
  type: vsphere
```

  url: <api_end_point> <1>
  settings:

>     vddkInitImage: <registry_route_or_server_path>/vddk:<tag> <2>

Reviewer: in 2.6 we add to the `settings` also the `sdkEndpoint` key that can be set with vcenter or esxi

  secret:
    name: <secret> <3>
    namespace: <namespace>
EOF
----
<1> Specify the API end point URL, for example, `https://<vCenter_host>/sdk`.

> <2> Specify the VDDK image that you created.

Reviewer: we should say that vddkInitImage is optional but point out that it's recommended

kind: Provider
metadata:
  name: <source_provider>
  namespace: <namespace>
spec:
  type: ova
  url:  <nfs_server:/nfs_path> <1>
  secret:
    name: <secret> <2>
    namespace: <namespace>
EOF
----
<1>  where: `nfs_server` is an IP or hostname of the server where the share was created and `nfs_path` is the path on the server where the OVA files are stored.
<2> Specify the name of provider `Secret` CR.
endif::[]

ifdef::ostack[]
[source,yaml,subs="attributes+"]
----
$ cat << EOF | {oc} apply -f -
apiVersion: forklift.konveyor.io/v1beta1
kind: Provider
metadata:
  name: <source_provider>
  namespace: <namespace>
spec:
  type: openstack
  url: <api_end_point> <1>
  secret:
    name: <secret> <2>
    namespace: <namespace>
EOF
----
<1> Specify the API end point URL, for example, `https://<identity_service>/v3`.
<2> Specify the name of provider `Secret` CR.
endif::[]

ifdef::cnv[]
[source,yaml,subs="attributes+"]
----
$ cat << EOF | {oc} apply -f -
apiVersion: forklift.konveyor.io/v1beta1
kind: Provider
metadata:
  name: <source_provider>
  namespace: <namespace>
spec:
  type: openshift
  url: <api_end_point> <1>
  secret:
    name: <secret> <2>
    namespace: <namespace>
EOF
----

> <1> Specify the API end point URL.

Reviewer: if we changed it in the secret then also change here to API server

+
[source,yaml,subs="attributes+"]
----
$  cat << EOF | {oc} apply -f -
apiVersion: forklift.konveyor.io/v1beta1
kind: Hook
metadata:
  name: <hook>
  namespace: <namespace>
spec:
  image: quay.io/konveyor/hook-runner
  playbook: |
    LS0tCi0gbmFtZTogTWFpbgogIGhvc3RzOiBsb2NhbGhvc3QKICB0YXNrczoKICAtIG5hbWU6IExv
    YWQgUGxhbgogICAgaW5jbHVkZV92YXJzOgogICAgICBmaWxlOiAiL3RtcC9ob29rL3BsYW4ueW1s
    IgogICAgICBuYW1lOiBwbGFuCiAgLSBuYW1lOiBMb2FkIFdvcmtsb2FkCiAgICBpbmNsdWRlX3Zh
    cnM6CiAgICAgIGZpbGU6ICIvdG1wL2hvb2svd29ya2xvYWQueW1sIgogICAgICBuYW1lOiB3b3Jr
    bG9hZAoK
EOF
----
+
where:
+
`playbook` refers to an optional Base64-encoded Ansible playbook. If you specify a playbook, the `image` must be `hook-runner`.
+
[NOTE]
====
You can use the default `hook-runner` image or specify a custom image. If you specify a custom image, you do not have to specify a playbook.
====
+

. Create a `Plan` manifest for the migration:
+
ifdef::vmware[]
[source,yaml,subs="attributes+"]
----
$ cat << EOF | {oc} apply -f -
apiVersion: forklift.konveyor.io/v1beta1
kind: Plan
metadata:
  name: <plan> <1>
  namespace: <namespace>
spec:

>   warm: true <2>

Reviewer: yeah, it doesn't matter much but still I'd prefer to set it to `false` here - as 'warm' is not the recommended setting

<3> Specify only one network map and one storage map per plan.
<4> Specify a network mapping even if the VMs to be migrated are not assigned to a network. The mapping can be empty in this case.
<5> Specify the name of the `NetworkMap` CR.
<6> Specify a storage mapping even if the VMs to be migrated are not assigned with disk images. The mapping can be empty in this case.
<7> Specify the name of the `StorageMap` CR.
<8> You can use either the `id` _or_ the `name` parameter to specify the source VMs. +
<9> Specify the VMware VM MOR..
<10> Optional: You can specify up to two hooks for a VM. Each hook must run during a separate migration step.
<11> Specify the name of the `Hook` CR.
<12> Allowed values are `PreHook`, before the migration plan starts, or `PostHook`, after the migration is complete.
endif::[]

ifdef::rhv[]
[source,yaml,subs="attributes+"]
----
$ cat << EOF | {oc} apply -f -
apiVersion: forklift.konveyor.io/v1beta1
kind: Plan
metadata:
  name: <plan> <1>
  namespace: <namespace>

> spec:

Reviewer: we're going to have another field here `preserveClusterCpuModel`, see explanation [here](https://github.com/kubev2v/forklift/pull/688/files#diff-09e2c8040fb3d68c4ad0656289202f8f305f29f8973eef899ca8d5efc947cba4R47)

<3> Specify only one network map and one storage map per plan.
<4> Specify a network mapping even if the VMs to be migrated are not assigned to a network. The mapping can be empty in this case.
<5> Specify the name of the `NetworkMap` CR.
<6> Specify a storage mapping even if the VMs to be migrated are not assigned with disk images. The mapping can be empty in this case.
<7> Specify the name of the `StorageMap` CR.
<8> You can use either the `id` _or_ the `name` parameter to specify the source VMs.
<9> Specify the {rhv-short} VM UUID.
<10> Optional: You can specify up to two hooks for a VM. Each hook must run during a separate migration step.
<11> Specify the name of the `Hook` CR.
<12> Allowed values are `PreHook`, before the migration plan starts, or `PostHook`, after the migration is complete.
endif::[]

ifdef::ova[]
[source,yaml,subs="attributes+"]
----
$ cat << EOF | {oc} apply -f -
apiVersion: forklift.konveyor.io/v1beta1
kind: Plan
metadata:
  name: <plan> <1>
  namespace: <namespace>
spec:

>   warm: true <2>

Reviewer: we shouldn't have `warm` documented for OVA

<3> Specify only one network map and one storage map per plan.
<4> Specify a network mapping even if the VMs to be migrated are not assigned to a network. The mapping can be empty in this case.
<5> Specify the name of the `NetworkMap` CR.
<6> Specify a storage mapping even if the VMs to be migrated are not assigned with disk images. The mapping can be empty in this case.
<7> Specify the name of the `StorageMap` CR.
<8> You can use either the `id` _or_ the `name` parameter to specify the source VMs.
<9> Specify the TBD.
<10> Optional: You can specify up to two hooks for a VM. Each hook must run during a separate migration step.
<11> Specify the name of the `Hook` CR.
<12> Allowed values are `PreHook`, before the migration plan starts, or `PostHook`, after the migration is complete.
endif::[]

ifdef::ostack[]
[source,yaml,subs="attributes+"]
----
$ cat << EOF | {oc} apply -f -
apiVersion: forklift.konveyor.io/v1beta1
kind: Plan
metadata:
  name: <plan> <1>
  namespace: <namespace>
spec:

>   warm: true <2>

Reviewer: we shouldn't document `warm` for OpenStack

<3> Specify only one network map and one storage map per plan.
<4> Specify a network mapping even if the VMs to be migrated are not assigned to a network. The mapping can be empty in this case.
<5> Specify the name of the `NetworkMap` CR.
<6> Specify a storage mapping even if the VMs to be migrated are not assigned with disk images. The mapping can be empty in this case.
<7> Specify the name of the `StorageMap` CR.
<8> You can use either the `id` _or_ the `name` parameter to specify the source VMs.
<9> Specify the {osp} VM UUID.
<10> Optional: You can specify up to two hooks for a VM. Each hook must run during a separate migration step.
<11> Specify the name of the `Hook` CR.
<12> Allowed values are `PreHook`, before the migration plan starts, or `PostHook`, after the migration is complete.
endif::[]

ifdef::cnv[]
[source,yaml,subs="attributes+"]
----
$ cat << EOF | {oc} apply -f -
apiVersion: forklift.konveyor.io/v1beta1
kind: Plan
metadata:
  name: <plan> <1>
  namespace: <namespace>
spec:

>   warm: true <2>

Reviewer: we shouldn't document `warm` for CNV

endif::[]

.Procedure
. Create a `Secret` manifest for the source provider credentials:
+
ifdef::vmware[]
[source,yaml,subs="attributes+"]
----
$ cat << EOF | {oc} apply -f -
apiVersion: v1
kind: Secret
metadata:
  name: <secret>
  namespace: <namespace>
  ownerReferences: <1>
    - apiVersion: forklift.konveyor.io/v1beta1
      kind: Provider
      name: <provider_name>
      uid: <provider_uid>
  labels:
    createdForProviderType: vsphere
    createdForResourceType: providers
type: Opaque
stringData:
  user: <user> <2>
  password: <password> <3>
  insecureSkipVerify: <true/false> <4>
  cacert: | <5>
    <ca_certificate>
  url: <api_end_point> <6>
EOF
----
<1> The `ownerReferences` section is optional.
<2> Specify the vCenter user.
<3> Specify the user password.
<4> Specify `<true>` to skip certificate verification, specify `false` to verify the certificate. Defaults to `false` if not specified. Skipping certificate verification proceeds with an insecure migration and then the certificate is not required. Insecure migration means that the transferred data is sent over an insecure connection and potentially sensitive data could be exposed.

> <5> Enter the {manager} CA certificate unless it was replaced by a third-party certificate, in which case enter the {manager} Apache CA certificate. You can retrieve the {manager} CA certificate at https://<engine_host>/ovirt-engine/services/pki-resource?resource=ca-certificate&format=X509-PEM-CA.

Reviewer: as you wrote above, this description should change for vSphere provider
I've downloaded the certificate as VMware explain [here](https://kb.vmware.com/s/article/2108294) but I don't think we should get to this level of details, especially as we enable to fetch the certificate from our UI
I'd just say that it's the CA certificate for the server that serves the SDK (either vCenter or ESXi, depending on the configuration of the provider)

<6> Required only when `type` is `multus`. Here, `namespace` can either be specified using the namespace property or with a name built as follows: `<network_namespace>/<network_name>`.
endif::[]

ifdef::cnv[]
[source,yaml,subs="attributes+"]
----
$  cat << EOF | {oc} apply -f -
apiVersion: forklift.konveyor.io/v1beta1
kind: NetworkMap
metadata:
  name: <network_map>
  namespace: <namespace>
spec:
  map:
    - destination:
        name: <network_name>
        type: pod <1>
      source:
        id: <source_network_id> <2>
        name: <source_network_name>
    - destination:
        name: <network_attachment_definition> <3>
        namespace: <network_attachment_definition_namespace> <4>
        type: multus
      source:
        name: <network_attachment_definition> <5>
        namespace: <network_attachment_definition_namespace> <6>
        type: multus
  provider:
    source:
      name: <source_provider>
      namespace: <namespace>
    destination:
      name: <destination_provider>
      namespace: <namespace>
EOF
----
<1> Allowed values are `pod` and `multus`.
<2> You can use either the `id` _or_ the `name` parameter to specify the source network. For `id`, specify the TBD.

> <3> Specify a network attachment definition for each additional {virt} network.

Author: @bennyz @liranr23 What is TBD here?

Reviewer 2: I don't understand which provider this is referring to

Author: @bennyz OpenShift Virtualization as source.

Reviewer 2: Virtual Machine I guess?

<1> Allowed values are `ReadWriteOnce` and `ReadWriteMany`.
<2> Specify the {rhv-short} storage domain UUID. For example, `f2737930-b567-451a-9ceb-2887f6207009`.
endif::[]

ifdef::ova[]
[NOTE]
====
For OVA, the `StorageMap` can map only a single storage, which all the disks from the OVA are associated with, to a storage class at the destination. For this reason, the storage is referred to in the UI as "Dummy storage for source provider <provider_name>".
====
+
[source,yaml,subs="attributes+"]
----
$ cat << EOF | {oc} apply -f -
apiVersion: forklift.konveyor.io/v1beta1
kind: StorageMap
metadata:
  name: <storage_map>
  namespace: <namespace>
spec:
  map:
    - destination:
        storageClass: <storage_class>
        accessMode: <access_mode> <1>
      source:
        name: <source_datastore> <2>
    - destination:
        storageClass: <storage_class>
        accessMode: <access_mode>
      source:
        name: <source_datastore>
  provider:
    source:
      name: <source_provider>
      namespace: <namespace>
    destination:
      name: <destination_provider>
      namespace: <namespace>
EOF
----
<1> Allowed values are `ReadWriteOnce` and `ReadWriteMany`.
<2> Specify the TBD.

> endif::[]

Author: @bkhizgiy What is TBD here?

Reviewer 3: @RichardHoch I'm not sure actually, seams to be sort of a place holder:/

Author: @ahadas What do you think?

Author: @ahadas 21 February -- suggestions? comment?

Author: @ahadas Just to be clear... if I write  name: "Dummy storage for source provider <provider_name>" , then the user will include the quotation marks in the YAML. Is that what we want?

Reviewer: @RichardHoch you're right, no need for the quotes but just to make it clear that `<provider_name>` should be replaced with the name of the provider

        accessMode: <access_mode>
      source:
        id: <source_datastore>
  provider:
    source:
      name: <source_provider>
      namespace: <namespace>
    destination:
      name: <destination_provider>
      namespace: <namespace>
EOF
----
<1> Allowed values are `ReadWriteOnce` and `ReadWriteMany`.
<2> Specify the {osp} `volume_type` UUID. For example, `f2737930-b567-451a-9ceb-2887f6207009`.
endif::[]

ifdef::cnv[]
[source,yaml,subs="attributes+"]
----
$ cat << EOF | {oc} apply -f -
apiVersion: forklift.konveyor.io/v1beta1
kind: StorageMap
metadata:
  name: <storage_map>
  namespace: <namespace>
spec:
  map:
    - destination:
        storageClass: <storage_class>
        accessMode: <access_mode> <1>
      source:
        id: <source_datastore> <2>
    - destination:
        storageClass: <storage_class>
        accessMode: <access_mode>
      source:
        id: <source_datastore>
  provider:
    source:
      name: <source_provider>
      namespace: <namespace>
    destination:
      name: <destination_provider>
      namespace: <namespace>
EOF
----
<1> Allowed values are `ReadWriteOnce` and `ReadWriteMany`.
<2> Specify the TBD.

> endif::[]

Author: @bennyz @liranr23 What is TBD here?

Reviewer 2: name is used for the source, not id, and the storage class name is set there

      namespace: <namespace>
  map: <3>
    network: <4>
      name: <network_map> <5>
      namespace: <namespace>
    storage: <6>
      name: <storage_map> <7>
      namespace: <namespace>
  targetNamespace: <target_namespace>
  vms: <8>
    - id: <source_vm> <9>
    - name: <source_vm>
      hooks: <10>
        - hook:
            namespace: <namespace>
            name: <hook> <11>
          step: <step> <12>
EOF
----
<1> Specify the name of the `Plan` CR.
<2> Specify whether the migration is warm or cold. If you specify a warm migration without specifying a value for the `cutover` parameter in the `Migration` manifest, only the precopy stage will run.
<3> Specify only one network map and one storage map per plan.
<4> Specify a network mapping even if the VMs to be migrated are not assigned to a network. The mapping can be empty in this case.
<5> Specify the name of the `NetworkMap` CR.
<6> Specify a storage mapping even if the VMs to be migrated are not assigned with disk images. The mapping can be empty in this case.
<7> Specify the name of the `StorageMap` CR.
<8> You can use either the `id` _or_ the `name` parameter to specify the source VMs.
<9> Specify the TBD.

> <10> Optional: You can specify up to two hooks for a VM. Each hook must run during a separate migration step.

Author: @bkhizgiy What is TBD here?

Author: @bkhizgiy Please help here...

Reviewer: @RichardHoch where's the TBD?

Reviewer: ahh you refer to the line above (920), I missed the TBD there somehow
@bkhizgiy what's the value of this field in a migration from OVA?..

  map: <3>
    network: <4>
      name: <network_map> <5>
      namespace: <namespace>
    storage: <6>
      name: <storage_map> <7>
      namespace: <namespace>
  targetNamespace: <target_namespace>
  vms: <8>
    - id: <source_vm> <9>
    - name: <source_vm>
      namespace: <namespace>
      hooks: <10>
        - hook:
            namespace: <namespace>
            name: <hook> <11>
          step: <step> <12>
EOF
----
<1> Specify the name of the `Plan` CR.
<2> Specify whether the migration is warm or cold. If you specify a warm migration without specifying a value for the `cutover` parameter in the `Migration` manifest, only the precopy stage will run.
<3> Specify only one network map and one storage map per plan.
<4> Specify a network mapping even if the VMs to be migrated are not assigned to a network. The mapping can be empty in this case.
<5> Specify the name of the `NetworkMap` CR.
<6> Specify a storage mapping even if the VMs to be migrated are not assigned with disk images. The mapping can be empty in this case.
<7> Specify the name of the `StorageMap` CR.
<8> You can use only the `name` parameter, not the `id` parameter, to specify the source VMs.
<9> Specify the TBD.

> <10> Optional: You can specify up to two hooks for a VM. Each hook must run during a separate migration step.

Author: @bennyz @liranr23 What is TBD here?

Reviewer 2: name and namespace should be used, id is not used

// Module included in the following assemblies:
//
// * documentation/doc-Migration_Toolkit_for_Virtualization/master.adoc

:_content-type: PROCEDURE
[id="new-migrating-virtual-machines-cli_{context}"]

ifdef::vmware[]
= Migrating a VMware vSphere source provider

Before you can migrate a VMware vSphere source provider using the CLI, you must obtain the SHA-1 fingerprint of a vCenter host, as described in the procedure that follows.

endif::[]
ifdef::rhv[]
= Migrating a {rhv-full} source provider

You can migrate a {rhv-full} ({rhv-short}) source provider using the CLI.

.Prerequisites

 If you are migrating a virtual machine with a direct LUN disk, ensure that the nodes in the {virt} destination cluster that the VM is expected to run on can access the backend storage.

> include::snip-migrating-luns.adoc[]

Author: @liranr23 Thanks for the heads up.

> include::snip-migrating-luns.adoc[]

Reviewer: right, should be covered by https://issues.redhat.com/browse/MTV-806

> include::snip-migrating-luns.adoc[]

Author: @ahadas This Jira https://issues.redhat.com/browse/MTV-659 is still "Testing"  has it been verified?

> include::snip-migrating-luns.adoc[]

Reviewer: as far as I know it wasn't verified yet as reflected by the status of the issue but I'll let @qiyuann keep me honest here

> include::snip-migrating-luns.adoc[]

Reviewer 4: verified now

> include::snip-migrating-luns.adoc[]

Author: @liranr23 Great -- the change to the LUNs snip is covered in https://github.com/kubev2v/forklift-documentation/pull/473.  Fibre Channel was only part of the note (snip .adoc file), so the call for it here is unaffected.

> include::snip-migrating-luns.adoc[]

Author: @liranr23 Is that the only issue you have left with this PR? Thanks.

// Module included in the following assemblies:
//
// * documentation/doc-Migration_Toolkit_for_Virtualization/master.adoc

:_content-type: PROCEDURE
[id="new-migrating-virtual-machines-cli_{context}"]

ifdef::vmware[]

> = Migrating a VMware vSphere source provider

Author: Doh! :)

You can migrate a VMware vSphere source provider using the CLI.

endif::[]
ifdef::rhv[]
= Migrating a {rhv-full} source provider

You can migrate a {rhv-full} ({rhv-short}) source provider using the CLI.

.Prerequisites

If you are migrating a virtual machine with a direct LUN disk, ensure that the nodes in the {virt} destination cluster that the VM is expected to run on can access the backend storage.

include::snip-migrating-luns.adoc[]

endif::[]
ifdef::ova[]
= Migrating an Open Virtual Appliance (OVA) source provider

You can migrate Open Virtual Appliance (OVA) files that were created by VMware vSphere as a source provider using the CLI.

> include::snippet_ova_tech_preview.adoc[]

Reviewer: in 2.6 the OVA provider wouldn't be tech preview anymore, would you like to drop it here or in a separate PR?

endif::[]
ifdef::ostack[]
= Migrating an {osp} source provider

You can migrate an {osp} source provider using the CLI.

[NOTE]
====
Migration using {osp} source providers only supports VMs that use only Cinder volumes.
====
endif::[]
ifdef::cnv[]
= Migrating a Red Hat {virt} source provider

You can use a Red Hat {virt} provider as both a source provider and destination provider.

> Specifically, the host cluster that is automatically added as a {virt} provider can be used as both a source provider and a destination provider.

Author: How's this?

ifdef::cnv[]
= Migrating from a Red Hat {virt} source provider

You can use a Red Hat {virt} provider as a source provider as well as a destination provider.

You can use the CLI to migrate VMs from the cluster that {project-short} is deployed on to another cluster, or from a remote cluster to the cluster that {project-short} is deployed on.
endif::[]

You can use the CLI to migrate VMs from the cluster that {project-short} is deployed on to another cluster, or from a remote cluster to the cluster that {project-short} is deployed on.
endif::[]

.Procedure
. Create a `Secret` manifest for the source provider credentials:
+
ifdef::vmware[]
[source,yaml,subs="attributes+"]
----
$ cat << EOF | {oc} apply -f -
apiVersion: v1
kind: Secret
metadata:
  name: <secret>
  namespace: <namespace>
  ownerReferences: <1>
    - apiVersion: forklift.konveyor.io/v1beta1
      kind: Provider
      name: <provider_name>
      uid: <provider_uid>
  labels:
    createdForProviderType: vsphere
    createdForResourceType: providers
type: Opaque
stringData:
  user: <user> <2>
  password: <password> <3>
  insecureSkipVerify: <true/false> <4>
  cacert: | <5>
    <ca_certificate>
  url: <api_end_point> <6>
EOF
----
<1> The `ownerReferences` section is optional.

> <2> Specify the vCenter user.

Reviewer: this would need to include the ESXi user (and below the same for password) that you added in a separate PR

Author: Done.

<3> Specify the user password.
<4> Specify `<true>` to skip certificate verification, specify `false` to verify the certificate. Defaults to `false` if not specified. Skipping certificate verification proceeds with an insecure migration and then the certificate is not required. Insecure migration means that the transferred data is sent over an insecure connection and potentially sensitive data could be exposed.

> <5> Enter the {manager} CA certificate unless it was replaced by a third-party certificate, in which case enter the {manager} Apache CA certificate. You can retrieve the {manager} CA certificate at https://<engine_host>/ovirt-engine/services/pki-resource?resource=ca-certificate&format=X509-PEM-CA.

Reviewer 5: ```suggestion
<5> Enter the {manager} CA certificate unless it was replaced by a third-party certificate, in which case. enter the {manager} Apache CA certificate. You can retrieve the {manager} CA certificate at https://<engine_host>/ovirt-engine/services/pki-resource?resource=ca-certificate&format=X509-PEM-CA.
```

> <5> Enter the CA certificate for the server that serves the SDK, either vCenter or ESXi, depending on the configuration of the provider, unless the CA certificate was replaced by a third-party certificate, in which case, enter the {manager} Apache CA certificate. You can retrieve the {manager} CA certificate at https://<engine_host>/ovirt-engine/services/pki-resource?resource=ca-certificate&format=X509-PEM-CA.

Reviewer: the identification of the server - great
the other part about custom CA certificate can be dropped
```suggestion
<5> Enter the CA certificate for the server that serves the SDK, either vCenter or ESXi, depending on the configuration of the provider.
```

Author: @ahadas , should i remove this for all providers or just for vSphere/ESXi?

> <6> Specify the API end point URL, for example, `https://<vCenter_host>/sdk`.

Reviewer: ```suggestion
<6> Specify the API endpoint URL, for example, `https://<vCenter_host>/sdk`.
```
I think that in our context it's `endpoint` as a single word (also applies to other providers below)
and this would also need to change to include ESXi endpoint

endif::[]
ifdef::rhv[]
[source,yaml,subs="attributes+"]
----
$ cat << EOF | {oc} apply -f -
apiVersion: v1
kind: Secret
metadata:
  name: <secret>
  namespace: <namespace>
  ownerReferences: <1>
    - apiVersion: forklift.konveyor.io/v1beta1
      kind: Provider
      name: <provider_name>
      uid: <provider_uid>
  labels:
    createdForProviderType: ovirt
    createdForResourceType: providers
type: Opaque
stringData:
  user: <user> <2>
  password: <password> <3>
  insecureSkipVerify: <true/false> <4>
  cacert: | <5>
    <ca_certificate>
  url: <api_end_point> <6>
EOF
----
<1> The `ownerReferences` section is optional.
<2> Specify the {rhv-short} {manager} user.
<3> Specify the user password.
<4> Specify `<true>` to skip certificate verification, specify `false` to verify the certificate. Defaults to `false` if not specified. Skipping certificate verification proceeds with an insecure migration and then the certificate is not required. Insecure migration means that the transferred data is sent over an insecure connection and potentially sensitive data could be exposed.

> <5> Enter the {manager} CA certificate unless it was replaced by a third-party certificate, in which case enter the {manager} Apache CA certificate. You can retrieve the {manager} CA certificate at https://<engine_host>/ovirt-engine/services/pki-resource?resource=ca-certificate&format=X509-PEM-CA.

Reviewer 5: ```suggestion
<5> Enter the {manager} CA certificate unless it was replaced by a third-party certificate, in which case, enter the {manager} Apache CA certificate. You can retrieve the {manager} CA certificate at https://<engine_host>/ovirt-engine/services/pki-resource?resource=ca-certificate&format=X509-PEM-CA.
```

> <5> Enter the CA certificate for the API server of the OpenShift cluster, unless it was replaced by a third-party certificate, in which case, enter the {manager} Apache CA certificate. You can retrieve the {manager} CA certificate at https://<engine_host>/ovirt-engine/services/pki-resource?resource=ca-certificate&format=X509-PEM-CA.

Reviewer: I think you changed the description of the server that is mentioned here by mistake - it shouldn't be `of the OpenShift cluster` in case of RHV

Author: Fixed.

endif::[]
ifdef::ostack[]
[source,yaml,subs="attributes+"]
----
$ cat << EOF | {oc} apply -f -
apiVersion: v1
kind: Secret
metadata:
  name: <secret>
  namespace: <namespace>
  ownerReferences: <1>
    - apiVersion: forklift.konveyor.io/v1beta1
      kind: Provider
      name: <provider_name>
      uid: <provider_uid>
  labels:
    createdForProviderType: openstack
    createdForResourceType: providers
type: Opaque
stringData:
  user: <user> <2>
  password: <password> <3>
  insecureSkipVerify: <true/false> <4>
  domainName: <domain_name>
  projectName: <project_name>
  regionName: <region name> <5>
  cacert: | <6>
    <ca_certificate>
  url: <api_end_point> <7>
EOF
----
<1> The `ownerReferences` section is optional.
<2> Specify the {osp} user.
<3> Specify the user password.
<4> Specify `<true>` to skip certificate verification, specify `false` to verify the certificate. Defaults to `false` if not specified. Skipping certificate verification proceeds with an insecure migration and then the certificate is not required. Insecure migration means that the transferred data is sent over an insecure connection and potentially sensitive data could be exposed.

> <5> Specify the name of the {osp} region.

Reviewer: why do we document the `regionName` this way and not `domainName` and `projectName`?

Author: @ahadas You're right. I have no idea why it's this way. Should I use callouts for all three or for none? I'd recommend none, since it seems trivial, but your call.

Reviewer: yeah, that's self explanatory, we can go with none

> <6> Enter the CA certificate for connecting to the source environment. The certificate is not used when `insecureSkipVerify` is set to `<true>`. Note that if you set `insecureSkipVerify` to `false` and you do not set  `cacert`, {project-short} attempts to use the system CA.

Reviewer: `The certificate is not used when insecureSkipVerify is set to <true>` - this also applies to the cacert property for other providers and we don't mention this above, do we really need this?

`Note that if you set insecureSkipVerify to false and you do not set  cacert, {project-short} attempts to use the system CA.` - this also applies to other providers and we don't mention it above, how about adding something like: `When this field is not set and 'skip certificate verification' is disabled, {project-short} attempts to use the system CA`, to all place where the cacert property appears?

Author: @RichardHoch start here.

Author: @ahadas I replaced all explanations of `cacert`, including vSphere and RHV, with "When this field is not set and 'skip certificate verification' is disabled, {project-short} attempts to use the system CA." Let me know if you wanted those two to be different, or if i otherwise misunderstood.

<7> Specify the API end point URL, for example, `https://<identity_service>/v3`.
endif::[]

ifdef::cnv[]
[source,yaml,subs="attributes+"]
----
$ cat << EOF | {oc} apply -f -
apiVersion: v1
kind: Secret
metadata:
  name: <secret>
  namespace: <namespace>
  ownerReferences: <1>
    - apiVersion: forklift.konveyor.io/v1beta1
      kind: Provider
      name: <provider_name>
      uid: <provider_uid>
  labels:
    createdForProviderType: openshift
    createdForResourceType: providers
type: Opaque
stringData:
  user: <user> <2>
  password: <password> <3>
  insecureSkipVerify: <true/false> <4>
  cacert: | <5>
    <ca_certificate>
  url: <api_end_point> <6>
EOF
----
<1> The `ownerReferences` section is optional.
<2> Specify the {virt} user.
<3> Specify the user password.
<4> Specify `<true>` to skip certificate verification, specify `false` to verify the certificate. Defaults to `false` if not specified. Skipping certificate verification proceeds with an insecure migration and then the certificate is not required. Insecure migration means that the transferred data is sent over an insecure connection and potentially sensitive data could be exposed.

> <5> Enter the {manager} CA certificate unless it was replaced by a third-party certificate, in which case, enter the {manager} Apache CA certificate. You can retrieve the {manager} CA certificate at https://<engine_host>/ovirt-engine/services/pki-resource?resource=ca-certificate&format=X509-PEM-CA.

Reviewer: ok, this confirms that there this part documents the cacert for RHV and vice versa

<6> Specify the URL of the end point of the API server.
endif::[]

[start=2]
. Create a `Provider` manifest for the source provider:
+
ifdef::vmware[]
[source,yaml,subs="attributes+"]
----
$ cat << EOF | {oc} apply -f -
apiVersion: forklift.konveyor.io/v1beta1
kind: Provider
metadata:
  name: <source_provider>
  namespace: <namespace>
spec:
  type: vsphere
  url: <api_end_point> <1>
  settings:

>     vddkInitImage: <registry_route_or_server_path>/vddk:<tag> <2>

Reviewer: this suggests that the image should always be named `vddk` which is incorrect - maybe we should replace the entire value with how images are documented in OpenShift?
```suggestion
    vddkInitImage: <how images are documented in OpenShift> <2>
```

Author: Done, with the explanation in callout.

    sdkEndpoint: <3>
  secret:
    name: <secret> <4>
    namespace: <namespace>
EOF
----
<1> Specify the API end point URL, for example, `https://<vCenter_host>/sdk`.

> <2> Optional, but strongly recommended to accelereatate migrations. Specify the VDDK image you created.

Reviewer 5: ```suggestion
<2> Optional, but it is strongly recommended to accelerate migrations. Specify the VDDK image you created.
```

apiVersion: forklift.konveyor.io/v1beta1
kind: Provider
metadata:
  name: <source_provider>
  namespace: <namespace>
spec:
  type: openstack
  url: <api_end_point> <1>
  secret:
    name: <secret> <2>
    namespace: <namespace>
EOF
----
<1> Specify the API end point URL, for example, `https://<identity_service>/v3`.
<2> Specify the name of provider `Secret` CR.
endif::[]

ifdef::cnv[]
[source,yaml,subs="attributes+"]
----
$ cat << EOF | {oc} apply -f -
apiVersion: forklift.konveyor.io/v1beta1
kind: Provider
metadata:
  name: <source_provider>
  namespace: <namespace>
spec:
  type: openshift
  url: <api_end_point> <1>
  secret:
    name: <secret> <2>
    namespace: <namespace>
EOF
----
<1> Specify the URL of the end point of the API server.
<2> Specify the name of provider `Secret` CR.
endif::[]

ifdef::vmware[]
[start=3]
. Create a `Host` manifest:
+
[source,yaml,subs="attributes+"]
----
$ cat << EOF | {oc} apply -f -
apiVersion: forklift.konveyor.io/v1beta1
kind: Host
metadata:
  name: <vmware_host>
  namespace: <namespace>
spec:
  provider:
    namespace: <namespace>
    name: <source_provider> <1>
  id: <source_host_mor> <2>
  ipAddress: <source_network_ip> <3>
EOF
----

> <1> Specify the name of the VMware `Provider` CR.

Author: Good catch! Went with VMware vSphere throughout the file. That's the correct way to document it -- and what we use throughout the rest of the document (I did not search all files -- I can do that later).

<2> Specify the managed object reference (MOR) of the VMware host.
<3> Specify the IP address of the VMware migration network.

[start=4]
. Create a `NetworkMap` manifest to map the source and destination networks:
+
[source,yaml,subs="attributes+"]
----
$  cat << EOF | {oc} apply -f -
apiVersion: forklift.konveyor.io/v1beta1
kind: NetworkMap
metadata:
  name: <network_map>
  namespace: <namespace>
spec:
  map:
    - destination:
        name: <network_name>
        type: pod <1>
      source:
        id: <source_network_id> <2>

>         name: <source_network_name>

Author: I moved the callout <2> so it's next to source. The callout text includes what we had in 2.4 and additional details about using `id`. Should I delete the additional details? [I think the answer is "no.']

>         name: <source_network_name>

Reviewer: > Should I delete the additional details? [I think the answer is "no.']

no, it's good to have the additional details

    - destination:
        name: <network_attachment_definition> <3>
        namespace: <network_attachment_definition_namespace> <4>
        type: multus

>       source:

Author: I went with replacing the source.

        name: <network_attachment_definition> <5>
        namespace: <network_attachment_definition_namespace> <6>
  provider:
    source:
      name: <source_provider>
      namespace: <namespace>
    destination:
      name: <destination_provider>
      namespace: <namespace>
EOF
----
<1> Allowed values are `pod` and `multus`.

> <2> You can use either the `id` _or_ the `name` parameter to specify the source network. For `id`, specify the VMware network MOR.

Author: @ahadas  I don't see a difference now that i moved the callout <2>

> <2> You can use either the `id` _or_ the `name` parameter to specify the source network. For `id`, specify the VMware network MOR.

Reviewer: yeah, me neither - it looks good

<3> Specify a network attachment definition for each additional {virt} network.
<4> Required only when `type` is `multus`. Specify the namespace of the {virt} network attachment definition.
<5> Specify a network attachment definition for each additional {virt} network.
<6> Required only when `type` is `multus`. Here, `namespace` can either be specified using the namespace property or with a name built as follows: `<network_namespace>/<network_name>`.
endif::[]

ifdef::rhv[]
[start=3]
. Create a `NetworkMap` manifest to map the source and destination networks:
+
[source,yaml,subs="attributes+"]
----
$  cat << EOF | {oc} apply -f -
apiVersion: forklift.konveyor.io/v1beta1
kind: NetworkMap
metadata:
  name: <network_map>
  namespace: <namespace>
spec:

>   map:

Author: same as above.

        id: <source_network_id> <2>
        name: <source_network_name>
    - destination:
        name: <network_attachment_definition> <3>
        namespace: <network_attachment_definition_namespace> <4>
        type: multus
      source:
        name: <network_attachment_definition> <5>
        namespace: <network_attachment_definition_namespace> <6>
  provider:
    source:
      name: <source_provider>
      namespace: <namespace>
    destination:
      name: <destination_provider>
      namespace: <namespace>
EOF
----
<1> Allowed values are `pod` and `multus`.
<2> You can use either the `id` _or_ the `name` parameter to specify the source network. For `id`, specify the {rhv-short} network UUID.
<3> Specify a network attachment definition for each additional {virt} network.
<4> Required only when `type` is `multus`. Specify the namespace of the {virt} network attachment definition.
<5> Specify a network attachment definition for each additional {virt} network.
<6> Required only when `type` is `multus`. Here, `namespace` can either be specified using the namespace property or with a name built as follows: `<network_namespace>/<network_name>`.
endif::[]

ifdef::ova[]
[start=3]

> . Create a `NetworkMap` manifest to map the source and destination networks:

Reviewer: same comments as above

Author: Likewise

kind: NetworkMap
metadata:
  name: <network_map>
  namespace: <namespace>
spec:
  map:
    - destination:
        name: <network_name>
        type: pod <1>
      source:
        id: <source_network_id>
    - destination:
        name: <network_attachment_definition> <2>
        namespace: <network_attachment_definition_namespace> <3>
        type: multus
      source:
        name: <network_attachment_definition> <4>
        namespace: <network_attachment_definition_namespace> <5>
  provider:
    source:
      name: <source_provider>
      namespace: <namespace>
    destination:
      name: <destination_provider>
      namespace: <namespace>
EOF
----
<1> Allowed values are `pod` and `multus`.
<2> Specify a network attachment definition for each additional {virt} network.
<3> Required only when `type` is `multus`. Specify the namespace of the {virt} network attachment definition.
<4> Specify a network attachment definition for each additional {virt} network.
<5> Required only when `type` is `multus`. Here, `namespace` can either be specified using the namespace property or with a name built as follows: `<network_namespace>/<network_name>`.
endif::[]

ifdef::ostack[]
[start=3]

> . Create a `NetworkMap` manifest to map the source and destination networks:

Reviewer: same comments as above (the source.id should refer to the OpenStack network UUID)

Author: Likewise,

        namespace: <network_attachment_definition_namespace> <6>
  provider:
    source:
      name: <source_provider>
      namespace: <namespace>
    destination:
      name: <destination_provider>
      namespace: <namespace>
EOF
----
<1> Allowed values are `pod` and `multus`.
<2> You can use either the `id` _or_ the `name` parameter to specify the source network. For `id`, specify the {osp} network UUID.
<3> Specify a network attachment definition for each additional {virt} network.
<4> Required only when `type` is `multus`. Specify the namespace of the {virt} network attachment definition.
<5> Specify a network attachment definition for each additional {virt} network.
<6> Required only when `type` is `multus`. Here, `namespace` can either be specified using the namespace property or with a name built as follows: `<network_namespace>/<network_name>`.
endif::[]

ifdef::cnv[]
[start=3]
. Create a `NetworkMap` manifest to map the source and destination networks:
+
[source,yaml,subs="attributes+"]
----
$  cat << EOF | {oc} apply -f -
apiVersion: forklift.konveyor.io/v1beta1
kind: NetworkMap
metadata:
  name: <network_map>
  namespace: <namespace>
spec:

>   map:

Reviewer: here we should keep the second mapping as is but to replace the source in the first mapping - should be a pod network that is similar to the destination

      namespace: <namespace>
    destination:
      name: <destination_provider>
      namespace: <namespace>
EOF
----
<1> Allowed values are `pod` and `multus`.
<2> You can use either the `id` _or_ the `name` parameter to specify the source network. For `id`, specify the virtual machine.
<3> Specify a network attachment definition for each additional {virt} network.
<4> Required only when `type` is `multus`. Specify the namespace of the {virt} network attachment definition.
<5> Specify a network attachment definition for each additional {virt} network.
<6> Required only when `type` is `multus`. Here, `namespace` can either be specified using the namespace property or with a name built as follows: `<network_namespace>/<network_name>`.
endif::[]

ifdef::vmware[]
[start=5]
. Create a `StorageMap` manifest to map source and destination storage:
+
[source,yaml,subs="attributes+"]
----
$ cat << EOF | {oc} apply -f -
apiVersion: forklift.konveyor.io/v1beta1
kind: StorageMap
metadata:
  name: <storage_map>
  namespace: <namespace>
spec:
  map:
    - destination:
        storageClass: <storage_class>
        accessMode: <access_mode> <1>
      source:
        id: <source_datastore> <2>

>     - destination:

Reviewer: the second mapping seems redundant (since it's exactly the same as the first one)
(also applies to the storage mappings for the other providers below)

kind: StorageMap
metadata:
  name: <storage_map>
  namespace: <namespace>
spec:
  map:
    - destination:
        storageClass: <storage_class>
        accessMode: <access_mode> <1>
      source:
        id: <source_datastore> <2>
    - destination:
        storageClass: <storage_class>
        accessMode: <access_mode>
      source:
        id: <source_datastore>
  provider:
    source:
      name: <source_provider>
      namespace: <namespace>
    destination:
      name: <destination_provider>
      namespace: <namespace>
EOF
----
<1> Allowed values are `ReadWriteOnce` and `ReadWriteMany`.
<2> Specify the {rhv-short} storage domain UUID. For example, `f2737930-b567-451a-9ceb-2887f6207009`.
endif::[]

ifdef::ova[]
[start=4]
. Create a `StorageMap` manifest to map source and destination storage:
+
[source,yaml,subs="attributes+"]
----
$ cat << EOF | {oc} apply -f -
apiVersion: forklift.konveyor.io/v1beta1
kind: StorageMap
metadata:
  name: <storage_map>
  namespace: <namespace>
spec:
  map:
    - destination:
        storageClass: <storage_class>
        accessMode: <access_mode> <1>
      source:
        name:  Dummy storage for source provider <provider_name> <2>

>     - destination:

Reviewer: here we really need to drop the second mapping since we have only a single "Dummy" storage for an OVA provider

----
$ cat << EOF | {oc} apply -f -
apiVersion: forklift.konveyor.io/v1beta1
kind: StorageMap
metadata:
  name: <storage_map>
  namespace: <namespace>
spec:
  map:
    - destination:
        storageClass: <storage_class>
        accessMode: <access_mode> <1>
      source:
        id: <source_datastore> <2>
    - destination:
        storageClass: <storage_class>
        accessMode: <access_mode>
      source:
        id: <source_datastore>
  provider:
    source:
      name: <source_provider>
      namespace: <namespace>
    destination:
      name: <destination_provider>
      namespace: <namespace>
EOF
----
<1> Allowed values are `ReadWriteOnce` and `ReadWriteMany`.
<2> Specify the {osp} `volume_type` UUID. For example, `f2737930-b567-451a-9ceb-2887f6207009`.
endif::[]

ifdef::cnv[]
[start=4]
. Create a `StorageMap` manifest to map source and destination storage:
+
[source,yaml,subs="attributes+"]
----
$ cat << EOF | {oc} apply -f -
apiVersion: forklift.konveyor.io/v1beta1
kind: StorageMap
metadata:
  name: <storage_map>
  namespace: <namespace>
spec:
  map:
    - destination:
        storageClass: <storage_class>
        accessMode: <access_mode> <1>
      source:

>         name: <source_class>

Reviewer: ```suggestion
        name: <storage_class>
```
?

<10> Optional: You can specify up to two hooks for a VM. Each hook must run during a separate migration step.
<11> Specify the name of the `Hook` CR.
<12> Allowed values are `PreHook`, before the migration plan starts, or `PostHook`, after the migration is complete.
endif::[]

ifdef::rhv[]
[start=6]
. Create a `Plan` manifest for the migration:
+
[source,yaml,subs="attributes+"]
----
$ cat << EOF | {oc} apply -f -
apiVersion: forklift.konveyor.io/v1beta1
kind: Plan
metadata:
  name: <plan> <1>
  namespace: <namespace>
  preserveClusterCpuModel: true <2>
spec:
  warm: true <3>
  provider:
    source:
      name: <source_provider>
      namespace: <namespace>
    destination:
      name: <destination_provider>
      namespace: <namespace>
  map: <4>
    network: <5>
      name: <network_map> <6>
      namespace: <namespace>
    storage: <7>
      name: <storage_map> <8>
      namespace: <namespace>
  targetNamespace: <target_namespace>
  vms: <9>
    - id: <source_vm> <10>
    - name: <source_vm>
      hooks: <11>
        - hook:
            namespace: <namespace>
            name: <hook> <12>
          step: <step> <13>
EOF
----
<1> Specify the name of the `Plan` CR.

> <2> Keeps stuff

Reviewer: this will be documented in a different PR?

Author: Added here.

<4> Specify only one network map and one storage map per plan.
<5> Specify a network mapping even if the VMs to be migrated are not assigned to a network. The mapping can be empty in this case.
<6> Specify the name of the `NetworkMap` CR.
<7> Specify a storage mapping even if the VMs to be migrated are not assigned with disk images. The mapping can be empty in this case.
<8> Specify the name of the `StorageMap` CR.
<9> You can use either the `id` _or_ the `name` parameter to specify the source VMs.
<10> Specify the {rhv-short} VM UUID.
<11> Optional: You can specify up to two hooks for a VM. Each hook must run during a separate migration step.
<12> Specify the name of the `Hook` CR.
<13> Allowed values are `PreHook`, before the migration plan starts, or `PostHook`, after the migration is complete.
endif::[]

ifdef::ova[]
[start=6]
. Create a `Plan` manifest for the migration:
+
[source,yaml,subs="attributes+"]
----
$ cat << EOF | {oc} apply -f -
apiVersion: forklift.konveyor.io/v1beta1
kind: Plan
metadata:
  name: <plan> <1>
  namespace: <namespace>
spec:

>   warm: false <2>

Reviewer: this property (warm) should be dropped for a plan from OVA

  provider:
    source:
      name: <source_provider>
      namespace: <namespace>
    destination:
      name: <destination_provider>
      namespace: <namespace>
  map: <3>
    network: <4>
      name: <network_map> <5>
      namespace: <namespace>
    storage: <6>
      name: <storage_map> <7>
      namespace: <namespace>
  targetNamespace: <target_namespace>
  vms: <8>
    - id: <source_vm> <9>
    - name: <source_vm>
      hooks: <10>
        - hook:
            namespace: <namespace>
            name: <hook> <11>
          step: <step> <12>
EOF
----
<1> Specify the name of the `Plan` CR.
<2> Specify whether the migration is warm - `true` - or cold - `false`. If you specify a warm migration without specifying a value for the `cutover` parameter in the `Migration` manifest, only the precopy stage will run.
<3> Specify only one network map and one storage map per plan.

> <4> Specify a network mapping even if the VMs to be migrated are not assigned to a network. The mapping can be empty in this case.

Reviewer 5: ```suggestion
<4> Specify a network mapping, even if the VMs to be migrated are not assigned to a network. The mapping can be empty in this case.
```

<5> Specify the name of the `NetworkMap` CR.
<6> Specify a storage mapping even if the VMs to be migrated are not assigned with disk images. The mapping can be empty in this case.
<7> Specify the name of the `StorageMap` CR.
<8> You can use either the `id` _or_ the `name` parameter to specify the source VMs.
<9> Specify the OVA VM UUID.
<10> Optional: You can specify up to two hooks for a VM. Each hook must run during a separate migration step.
<11> Specify the name of the `Hook` CR.
<12> Allowed values are `PreHook`, before the migration plan starts, or `PostHook`, after the migration is complete.
endif::[]

ifdef::ostack[]
[start=6]
. Create a `Plan` manifest for the migration:
+
[source,yaml,subs="attributes+"]
----
$ cat << EOF | {oc} apply -f -
apiVersion: forklift.konveyor.io/v1beta1
kind: Plan
metadata:
  name: <plan> <1>
  namespace: <namespace>
spec:

>   warm: false <2>

Reviewer: `warm` should be dropped from here as well

  provider:
    source:
      name: <source_provider>
      namespace: <namespace>
    destination:
      name: <destination_provider>
      namespace: <namespace>
  map: <3>
    network: <4>
      name: <network_map> <5>
      namespace: <namespace>
    storage: <6>
      name: <storage_map> <7>
      namespace: <namespace>
  targetNamespace: <target_namespace>
  vms: <8>
    - id: <source_vm> <9>
    - name: <source_vm>
      hooks: <10>
        - hook:
            namespace: <namespace>
            name: <hook> <11>
          step: <step> <12>
EOF
----
<1> Specify the name of the `Plan` CR.
<2> Specify whether the migration is warm - `true` - or cold - `false`. If you specify a warm migration without specifying a value for the `cutover` parameter in the `Migration` manifest, only the precopy stage will run.
<3> Specify only one network map and one storage map per plan.

> <4> Specify a network mapping even if the VMs to be migrated are not assigned to a network. The mapping can be empty in this case.

Reviewer 5: ```suggestion
<4> Specify a network mapping, even if the VMs to be migrated are not assigned to a network. The mapping can be empty in this case.
```

<5> Specify the name of the `NetworkMap` CR.
<6> Specify a storage mapping, even if the VMs to be migrated are not assigned with disk images. The mapping can be empty in this case.
<7> Specify the name of the `StorageMap` CR.
<8> You can use either the `id` _or_ the `name` parameter to specify the source VMs.
<9> Specify the {osp} VM UUID.
<10> Optional: You can specify up to two hooks for a VM. Each hook must run during a separate migration step.
<11> Specify the name of the `Hook` CR.
<12> Allowed values are `PreHook`, before the migration plan starts, or `PostHook`, after the migration is complete.
endif::[]

ifdef::cnv[]
[start=6]
. Create a `Plan` manifest for the migration:
+
[source,yaml,subs="attributes+"]
----
$ cat << EOF | {oc} apply -f -
apiVersion: forklift.konveyor.io/v1beta1
kind: Plan
metadata:
  name: <plan> <1>
  namespace: <namespace>
spec:

>   warm: false <2>

Reviewer: `warm` should also be dropped from here

  provider:
    source:
      name: <source_provider>
      namespace: <namespace>
    destination:
      name: <destination_provider>
      namespace: <namespace>
  map: <3>
    network: <4>
      name: <network_map> <5>
      namespace: <namespace>
    storage: <6>
      name: <storage_map> <7>
      namespace: <namespace>
  targetNamespace: <target_namespace>
  vms:
    - name: <source_vm>
      namespace: <namespace>
      hooks: <8>
        - hook:
            namespace: <namespace>
            name: <hook> <9>
          step: <step> <10>
EOF
----
<1> Specify the name of the `Plan` CR.
<2> Specify whether the migration is warm - `true` - or cold - `false`. If you specify a warm migration without specifying a value for the `cutover` parameter in the `Migration` manifest, only the precopy stage will run.
<3> Specify only one network map and one storage map per plan.

> <4> Specify a network mapping even if the VMs to be migrated are not assigned to a network. The mapping can be empty in this case.

Reviewer 5: ```suggestion
<4> Specify a network mapping, even if the VMs to be migrated are not assigned to a network. The mapping can be empty in this case.
```

<5> Specify the name of the `NetworkMap` CR.

> <6> Specify a storage mapping even if the VMs to be migrated are not assigned with disk images. The mapping can be empty in this case.

Reviewer 5: ```suggestion
<6> Specify a storage mapping, even if the VMs to be migrated are not assigned with disk images. The mapping can be empty in this case.
```

<7> Specify the name of the `StorageMap` CR.
<8> Optional: You can specify up to two hooks for a VM. Each hook must run during a separate migration step.
<9> Specify the name of the `Hook` CR.
<10> Allowed values are `PreHook`, before the migration plan starts, or `PostHook`, after the migration is complete.
endif::[]
+
. Create a `Migration` manifest to run the `Plan` CR:
+
[source,yaml,subs="attributes+"]
----
$ cat << EOF | {oc} apply -f -
apiVersion: forklift.konveyor.io/v1beta1
kind: Migration
metadata:
  name: <name_of_migration_cr>
  namespace: <namespace>
spec:
  plan:
    name: <name_of_plan_cr>
    namespace: <namespace>

>   cutover: <optional_cutover_time>

Reviewer: we can either duplicate the migration section per-provider and keep `cutover` only for vSphere and RHV or to document here that it's relevant only when `warm` is set to true

about the comments above on dropping the `warm` property for openstack, ova and openshift providers, we can alternatively keep it as document that it has no effect
